rust/src/ftp/ftp.rs

```diff
@@ -0,0 +1,209 @@
+        FtpCommand::new("IDLE", FTP_COMMAND_IDLE, 4),
+        FtpCommand::new("LIST", FTP_COMMAND_LIST, 4),
+        FtpCommand::new("MAIL", FTP_COMMAND_MAIL, 4),
+        FtpCommand::new("MDTM", FTP_COMMAND_MDTM, 4),
+        FtpCommand::new("MKD", FTP_COMMAND_MKD, 3),
+        FtpCommand::new("MLFL", FTP_COMMAND_MLFL, 4),
+        FtpCommand::new("MODE", FTP_COMMAND_MODE, 4),
+        FtpCommand::new("MRCP", FTP_COMMAND_MRCP, 4),
+        FtpCommand::new("MRSQ", FTP_COMMAND_MRSQ, 4),
+        FtpCommand::new("MSAM", FTP_COMMAND_MSAM, 4),
+        FtpCommand::new("MSND", FTP_COMMAND_MSND, 4),
+        FtpCommand::new("MSOM", FTP_COMMAND_MSOM, 4),
+        FtpCommand::new("NLST", FTP_COMMAND_NLST, 4),
+        FtpCommand::new("NOOP", FTP_COMMAND_NOOP, 4),
+        FtpCommand::new("PASS", FTP_COMMAND_PASS, 4),
+        FtpCommand::new("PWD", FTP_COMMAND_PWD, 3),
+        FtpCommand::new("QUIT", FTP_COMMAND_QUIT, 4),
+        FtpCommand::new("REIN", FTP_COMMAND_REIN, 4),
+        FtpCommand::new("REST", FTP_COMMAND_REST, 4),
+        FtpCommand::new("RMD", FTP_COMMAND_RMD, 3),
+        FtpCommand::new("RNFR", FTP_COMMAND_RNFR, 4),
+        FtpCommand::new("RNTO", FTP_COMMAND_RNTO, 4),
+        FtpCommand::new("SITE", FTP_COMMAND_SITE, 4),
+        FtpCommand::new("SIZE", FTP_COMMAND_SIZE, 4),
+        FtpCommand::new("SMNT", FTP_COMMAND_SMNT, 4),
+        FtpCommand::new("STAT", FTP_COMMAND_STAT, 4),
+        FtpCommand::new("STOU", FTP_COMMAND_STOU, 4),
+        FtpCommand::new("STRU", FTP_COMMAND_STRU, 4),
+        FtpCommand::new("SYST", FTP_COMMAND_SYST, 4),
+        FtpCommand::new("TYPE", FTP_COMMAND_TYPE, 4),
+        FtpCommand::new("UMASK", FTP_COMMAND_UMASK, 5),
+        FtpCommand::new("USER", FTP_COMMAND_USER, 4),
+        FtpCommand::new("UNKNOWN", FTP_COMMAND_UNKNOWN, 7),
+        FtpCommand::new("MAX", FTP_COMMAND_MAX, 0),
+    ];
+}
+
+#[allow(non_snake_case)]
+#[no_mangle]
+pub unsafe extern "C" fn SCGetFtpCommandInfo(
+    index: usize, name_ptr: *mut *const c_char, code_ptr: *mut u8, len_ptr: *mut u8,
+) -> bool {
+    if index <= FTP_COMMANDS.len() {
+        unsafe {
+            if !name_ptr.is_null() {
+                *name_ptr = FTP_COMMANDS[index].command_name.as_ptr();
+            }
+            if !code_ptr.is_null() {
+                *code_ptr = FTP_COMMANDS[index].command;
+            }
+            if !len_ptr.is_null() {
+                *len_ptr = FTP_COMMANDS[index].command_length;
+            }
+        }
+        true
+    } else {
+        false
+    }
+}
+
+#[allow(non_snake_case)]
+#[no_mangle]
+pub extern "C" fn SCGetFtpCommandTableSize() -> usize {
+    FTP_COMMANDS.len()
+}
+
+#[repr(C)]
+#[allow(dead_code)]
+pub struct FtpTransferCmd {
+    // Must be first -- required by app-layer expectation logic
+    data_free: unsafe extern "C" fn(*mut c_void),
+    pub flow_id: u64,
+    pub file_name: *mut u8,
+    pub file_len: u16,
+    pub direction: u8,
```

Comment:
little change in `c5fea52f4d126ebbb49ab5a04dcd82a94e9f4b95` can be squashed into `93b9950f374ece3c424db647f38bfb4f509b3d2d`